### DISCUSSION TITLE
CDK Custom Resources - feedback request

### DISCUSSION THREAD

**Discussion by evgenyka** [aws/aws-cdk#25194] [Q&A]
Hey CDK community!
We'd like to start a discussion here about issues that are blocking you from using CDK custom resources.
Custom resources are a powerful feature that allows you to define AWS resources and implement functionality that is not natively supported by AWS CloudFormation. The CDK also utilizes them internally in some constructs to provide enhanced capabilities, such as `BucketDeployment` and `KubernetesManifest`, to name a few.
We want to make sure that they are working as intended and meeting your needs. If you've encountered any issues or 
challenges with custom resources that are preventing you from leveraging them, or any CDK feature powered by them, 
we'd love to hear from you here.
If you prefer to have a private discussion, feel free to contact me directly.
Your feedback is essential to us, as we continue to improve and enhance the CDK experience for all users.
Thanks!

**Comment by mnapoli** [comment]
Custom resources are confusing because it creates Lambda function "polluting" the application/the account.

I wish it was possible to use custom resources without having these "deployment utility" functions in the same place as production functions.

**Reply by otaviomacedo** [reply]
Is it just an inconvenience, or is it an actual blocker in your case?

**Reply by mnapoli** [reply]
A blocker. I am not using the S3 option to automatically clean up the buckets for example, and instead I am staying on Serverless Framework and/or developing an alternative CLI to the AWS CDK to implement this behavior in the CLI itself.

**Comment by rittneje** [comment]
Custom resources are difficult to get right because of the nuances of how CloudFormation works. Even the CDK team gets it wrong. See for example #22670, #17443, and #16603.

**Comment by malikalimoekhamedov** [comment]
AWS Custom Resources is a significant but very obscure feature. The documentation doesn't take you by the hand; it sounds somewhat cryptic to someone who never played around with it. There are too few examples, and it's not immediately apparent why one would go with `AwsCustomResource` vs `CustomResource`. Most of the documentation uses CDK v1 to describe use cases. You turn to YouTube to find out more, only to find out paraphrasing of AWS blog posts on the topic.

It took me a very long time to get my head around it, and I'm still unable to get it to work (see #26657).

**Reply by rittneje** [reply]
CustomResource is the actual L2 construct for custom resources. It lets you define a lambda function to do whatever you need. You can get more details in the [CloudFormation docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources.html), but it certainly requires a lot of effort to implement it properly for all state transitions.

AwsCustomResource is intended as a convenience wrapper around CustomResource for when your custom resource is backed by simple SDK calls. Unfortunately, as I described in #17443, it is broken by design and thus unsafe for use in most cases.

**Reply by malikalimoekhamedov** [reply]
Thanks, @rittneje.

Yes, I did understand the difference between them though it took me a bit of time the first time I tried to get into custom resources.
You seem to have more experience and a deeper understanding of this part of CDK. Would you have an idea as to why I encountered issue #26657? Maybe you've experienced it before?
